File: Reqnroll/Chapter14/14.3-TheObjectMotherPattern/After/WIMP.Specs/Support/OrderObjectMother.cs
using WIMP.App.Models;

namespace WIMP.Specs.Support;

public class OrderObjectMother
{
    private readonly Order order = new();

    public OrderObjectMother()
    {
        // add default item
        order.AddItem(DomainDefaults.PizzaItemDefaultInstance());
        // add default address
        order.DeliveryAddress = DomainDefaults.DeliveryAddress;
    }

    public Order Build()
    {
        return order;
    }

    public OrderObjectMother WithAdditionalItem(
        string? name = null, string? size = null, PizzaStyle? style = null)
    {
        var pizzaItem = DomainDefaults.PizzaItemDefaultInstance(name, size, style);
        order.AddItem(pizzaItem);
        return this;
    }

    public OrderObjectMother WithItem(
        string? name = null, string? size = null, PizzaStyle? style = null)
    {
        order.ClearItems(); // remove existing items
        return WithAdditionalItem(name, size, style);
    }

    public OrderObjectMother WithItems(int quantity,
        string? name = null, string? size = null, PizzaStyle? style = null)
    {
        order.ClearItems(); // remove existing items
        for (int i = 0; i < quantity; i++)
        {
            WithAdditionalItem(name, size, style);
        }
        return this;
    }
}

File: CucumberJVM/Chapter14/14.3-ObjectMotherPattern/After/src/test/java/com/wimp/app/specs/support/OrderObjectMother.java
package com.wimp.app.specs.support;

import com.wimp.app.models.Order;
import com.wimp.app.models.PizzaItem;
import com.wimp.app.models.PizzaStyle;

public class OrderObjectMother {
    private final Order order;

    public OrderObjectMother() {
        order = new Order();
        // add default item
        order.addItem(DomainDefaults.pizzaItemDefaultInstance());
        order.setDeliveryAddress(DomainDefaults.CUSTOMER_ADDRESS);
    }

    public Order build() {
        return order;
    }

    public OrderObjectMother withAdditionalItem(String name, String size, PizzaStyle style) {
        PizzaItem pizzaItem = DomainDefaults.pizzaItemDefaultInstance(name, size, style);
        order.addItem(pizzaItem);
        return this;
    }

    public OrderObjectMother withItem(String name, String size, PizzaStyle style) {
        order.clearItems(); // remove existing items
        return withAdditionalItem(name, size, style);
    }

    public OrderObjectMother withItem(String name) {
        return withItem(name, null, null);
    }

    public OrderObjectMother withItems(int quantity, String name, String size, PizzaStyle style) {
        order.clearItems(); // remove existing items
        for (int i = 0; i < quantity; i++) {
            withAdditionalItem(name, size, style);
        }
        return this;
    }

    public OrderObjectMother withItems(int quantity, String size) {
        return withItems(quantity, null, size, null);
    }
}
